File: world.py
# imports
from asset_manager import get_asset
import pygame, random, terrain, decoration, aplayer, lighting, math, os, time, enemies, nest, laser, gateway, UI, loading_screen
from util import rotateAndGetOffset

class World:
    def __init__(self, worldWidth, worldHeight, loading_screen: loading_screen.LoadingScreen, defaultZooms=[0.1, 2], developingMode=False):
        self.worldWidth   = worldWidth
        self.worldHeight  = worldHeight
        self.defaultZooms = defaultZooms
        self.developingMode = developingMode

        init_loading_screen, objects_loading_screen, generate_loading_screen = loading_screen.subsections(0, 0.25, 0.5)

        inits = [lighting.init, enemies.init, nest.init, terrain.init, aplayer.init, laser.init, gateway.init, UI.init]

        for i, init in enumerate(inits):
            init_loading_screen.put((i + 1) / len(inits), f"{init.__module__}.{init.__name__}()")
            init()

        self.decorations = []

        objects_loading_screen.put(0.5, "Creating terrain object")
        self.terrain = terrain.Terrain(worldWidth, worldHeight, defaultZooms=defaultZooms)
        objects_loading_screen.put(0.6, "Creating player object")
        self.player = aplayer.Player(defaultZooms, worldWidth / 2, -200 if developingMode else -1200)
        objects_loading_screen.put(0.7, "Creating lighting object")
        self.light  = lighting.Lighting(defaultZooms=defaultZooms)
        objects_loading_screen.put(0.8, "Creating background surface")
        background_raw   = get_asset("Background")
        self.background  = pygame.transform.scale(background_raw, (4000, 4000))
        self.bg_width, self.bg_height = self.background.get_size()
        objects_loading_screen.put(0.9, "Creating foreground surface")
        foreground_raw   = get_asset("Foreground")
        self.foreground  = pygame.transform.scale(foreground_raw, (10000, 10000))
        self.fg_width, self.fg_height = self.foreground.get_size()
        objects_loading_screen.put(1.0, "Object creation complete.")

        self._world_layer      = None
        self._world_layer_size = None
        self.scratch_layer = None

        self.generateWorld(generate_loading_screen)

    def generateWorld(self, loading_screen):
        self.terrain.generateLayer(0, loading_screen)

    def generateNextLayer(self):
        self.terrain.generateLayer(1)

    def _getWorldLayer(self, real_window_size):
        if self._world_layer is None or self._world_layer_size != real_window_size:
            self._world_layer = pygame.Surface(real_window_size)
            self._world_layer_size = real_window_size
            self.scratch_layer = pygame.Surface(real_window_size)
        return self._world_layer, self.scratch_layer

    def addAirPocket(self, x, y, radius):
        # player-mined pockets go into the layer the player is currently in
        layerIndex = self.terrain._layerForY(self.player.y)
        self.terrain.addAirPocket(x, y, radius, layerIndex=layerIndex, playerMade=True)

    def healNests(self):
        for li in range(terrain.NUM_LAYERS):
            for n in self.terrain.nests[li]:
                if n.health > 0:
                    n.health = n.maxHealth
                    n.stage  = 0

    def removeEnemies(self):
        for li in range(terrain.NUM_LAYERS):
            for n in self.terrain.nests[li]:
                n.enemies.clear()

    def tick(self, FPS, window_size, frame, mousePos, keysDown, events):
        left, top, zoom = frame
        frameLength = 1000 / FPS

        # update which layers are active this frame
        self.terrain.updateActiveLayers(self.player.y)

        self.terrain.newKnockbackCircles    = []
        self.terrain.newPlayerDamageCircles = []

        if self.player.tick(frameLength, self.terrain, mousePos, keysDown, events):
            return True

        if random.randint(1, math.ceil(FPS / 7)) == 1:
            self.light.addMistParticle(self.player.x, self.player.y, color=self.player.color)
        for lase in self.player.laser:
            if random.randint(1, math.ceil(FPS / max(1, lase.length) * 25)) == 1:
                mistPos = random.random()
                self.light.addMistParticle(
                    lase.startX + mistPos * lase.length * math.cos(lase.angle),
                    lase.startY + mistPos * lase.length * math.sin(lase.angle),
                    color=self.player.color)

        w_width, w_height = window_size
        w_r = math.sqrt(w_width ** 2 + w_height ** 2) / 2 / zoom
        x, y = left + w_width / zoom / 2, top + w_height / zoom / 2

        # tick gateways
        for lase in self.player.laser:
            if self.terrain.playerDamageCircles and lase.collision:
                lx,ly=lase.collision[0]
                for gw in self.terrain.gateways:
                    if gw.tick(self.terrain, self.player, lx,ly):
                        self.terrain.particles.spawnMiningParticles(10, (255,255,255), 10, lx,ly)
                        break

        # active nests only
        for li in self.terrain.activeLayers:
            for n in self.terrain.nests[li]:
                n.updateVisuals(frameLength)
                if self.terrain.playerDamageCircles:
                    for particleCoords in n.applyDamageFromCircles(self.terrain, self.player):
                        self.terrain.particles.spawnMiningParticles(
                            10, n.color, particleCoords[2], particleCoords[0], particleCoords[1])

                if n.stage != n.maxStage:
                    ndx  = self.player.x - n.x
                    ndy  = self.player.y - n.y
                    d_sq = ndx * ndx + ndy * ndy
                    if d_sq < 300 * 300 and random.randint(1, int(200 + 0.1 * int(math.sqrt(d_sq) / 2) ** 2)) < frameLength:
                        n.addEnemy(self.terrain, self.player)
                    for i in range(len(n.enemies) - 1, -1, -1):
                        enemy = n.enemies[i]
                        if enemy.tick(frameLength, self.terrain, self.player):
                            del n.enemies[i]

                if random.randint(1, math.ceil(FPS / 6)) == 1 and n.close(x, y, w_r) and n.stage == n.maxStage:
                    self.light.addMistParticle(n.x, n.y, color=n.color)

        self.light.tickEffects(frameLength)
        self.terrain.particles.tickParticles(frameLength)

        self.terrain.knockbackCircles    = self.terrain.newKnockbackCircles
        self.terrain.playerDamageCircles = self.terrain.newPlayerDamageCircles

        return False

    def drawBackground(self, layer, window_size, frame):
        left, top, zoom = frame
        x = (-left * zoom) % self.bg_width/2  - self.bg_width/2
        y = (-top  * zoom) % self.bg_height/2 - self.bg_height/2
        layer.blit(self.background, (x, y))
    
    def drawForeground(self,layer,window_size,frame):
        left, top, zoom = frame
        x = (-left*6 * zoom) % self.fg_width/2  - self.fg_width/2
        y = ((-top*6+500)  * zoom) % self.fg_height/2 - self.fg_height/2
        layer.blit(self.foreground, (x, y))

    def getSurface(self, window_size, frame, hitboxes=False, kindVisibility=False,
                   real_window_size=None, offset_x=0, offset_y=0, tilt=0, crosshair=False):
        if real_window_size is None:
            real_window_size = window_size

        layer, scratchLayer = self._getWorldLayer(real_window_size)

        if kindVisibility:
            layer.fill((200, 200, 200))
        else:
            self.terrain.drawDepthBackground(layer, frame, offset_x=offset_x, offset_y=offset_y)

        

        self.light.drawEffects(layer, frame, offset_x=offset_x, offset_y=offset_y)

        self.light.drawGradient(layer, frame, self.player.color,
                                self.player.x, self.player.y,
                                offset_x=offset_x, offset_y=offset_y)
        if self.player.laser:
            if self.player.laser[0].collision:
                cx, cy = self.player.laser[0].collision[0]
                self.light.drawGradient(layer, frame, self.player.color, cx, cy,
                                        offset_x=offset_x, offset_y=offset_y)

        self.terrain.drawNestGradients(window_size, layer, frame,
                                       offset_x=offset_x, offset_y=offset_y)

        

        self.drawBackground(scratchLayer, window_size, frame)

        # gateway back elements (behind terrain)
        for gw in self.terrain.gateways:
            gw.drawBack(scratchLayer, frame, offset_x=offset_x, offset_y=offset_y)
        
        layer.blit(scratchLayer,(0,0),special_flags=pygame.BLEND_RGB_MULT)

        self.player.draw(layer, frame, hitboxes=hitboxes,
                         offset_x=offset_x, offset_y=offset_y, tilt=tilt)

        self.terrain.drawEnemies(window_size, layer, frame, hitboxes=hitboxes,
                                 offset_x=offset_x, offset_y=offset_y)

        self.terrain.particles.drawParticles(layer, frame,
                                             offset_x=offset_x, offset_y=offset_y)

        self.terrain.drawNests(window_size, layer, frame, hitboxes=hitboxes,
                               offset_x=offset_x, offset_y=offset_y)

        self.terrain.drawTerrain(window_size, layer, frame, hitboxes=hitboxes,
                                 real_window_size=real_window_size,
                                 offset_x=offset_x, offset_y=offset_y)

        # gateway front elements (after terrain)
        for gw in self.terrain.gateways:
            gw.draw(layer, frame, offset_x=offset_x, offset_y=offset_y)
        
        self.terrain.drawHealthBars(window_size, layer, frame, pygame.time.get_ticks(),offset_x=offset_x,offset_y=offset_y)

        self.drawForeground(scratchLayer, window_size, frame)
        self.light.drawThickGradient(scratchLayer,frame,self.player.x, self.player.y, offset_x=offset_x, offset_y=offset_y) 
        layer.blit(self.scratch_layer, (0,0), special_flags=pygame.BLEND_MULT)

        if crosshair:
            pygame.draw.line(layer, (100, 100, 100, 0.3), (real_window_size[0]*0.45, real_window_size[1]//2), (real_window_size[0]*0.55, real_window_size[1]//2), 2)
            pygame.draw.line(layer, (100, 100, 100, 0.3), (real_window_size[0]//2, real_window_size[1]*0.45), (real_window_size[0]//2, real_window_size[1]*0.55), 2)

        if tilt != 0:
            layer, cx, cy = rotateAndGetOffset(layer, real_window_size[0]/2, real_window_size[1]/2, math.radians(tilt))
            layer.blit(layer, (cx, cy))

        if crosshair:
            size = 10
            pygame.draw.line(layer, (255, 0, 0), (real_window_size[0]//2 - size, real_window_size[1]//2), (real_window_size[0]//2 + size, real_window_size[1]//2), 2)
            pygame.draw.line(layer, (255, 0, 0), (real_window_size[0]//2, real_window_size[1]//2 - size), (real_window_size[0]//2, real_window_size[1]//2 + size), 2)
                

        return layer

File: cpp_port/World.cpp
#include "World.h"
#include "Canvas.h"
#include "BlendModes.h"
#include "GlobalAssets.h"
#include "Util.h"
#include "Time.h"
#include "Cell.h"
#include "Enemy.h"
#include "Nest.h"
#include "InteractionDisplay.h"
#include <cmath>
#include <algorithm>

// was: inits = [lighting.init, cells.init, enemies.init, nest.init,
//               terrain.init, player.init, laser.init,
//               interaction_display.init, charge_display.init]
void World::init(SDL_Renderer* renderer, const std::string& interactionFontPath) {
    Lighting::init(renderer);
    Cell::init(renderer);
    Enemy::init(renderer);
    Nest::init(renderer);
    Terrain::init(renderer);
    Player::init(renderer);
    Laser::init();
    InteractionDisplay::init(renderer, interactionFontPath);
    // charge_display.init() skipped -- see header note.
}

World::World(SDL_Renderer* renderer, int worldWidth, int worldHeight, bool developingModeIn)
    : terrain(worldWidth, worldHeight),
      player(renderer, worldWidth / 2.0, developingModeIn ? -200.0 : -1200.0),
      developingMode(developingModeIn),
      renderer_(renderer) {
    generateWorld();
    terrain.startStreaming();
}

void World::generateWorld() {
    terrain.generateWorld(renderer_);
}

// def heal_nests(self): ...
void World::healNests() {
    for (auto& n : terrain.nests) {
        if (n->health > 0) {
            n->health = n->maxHealth;
            n->stage = 0;
        }
    }
}

// def remove_enemies(self): ...
void World::removeEnemies() {
    for (auto& n : terrain.nests) {
        n->enemies.clear();
    }
    terrain.enemies.clear();
}

// def tick(self, fps, window_size, frame, mouse_pos, keys_down, events): ...
bool World::tick(double fps, const Frame& frame, Vec2 mousePos, const PlayerInput& input,
                  const std::unordered_map<SDL_Keycode, bool>& keysDown) {
    double frameLength = 1000.0 / fps;
    Rect screenRect{
        static_cast<int>(frame.left), static_cast<int>(frame.top),
        static_cast<int>(frame.viewWidth / frame.zoom), static_cast<int>(frame.viewHeight / frame.zoom)
    };

    terrain.updateStreaming(player.x, player.y);

    terrain.newKnockbackCircles.clear();
    terrain.newPlayerDamageCircles.clear();

    if (player.tick(frameLength, terrain, mousePos, input)) return true;

    double playerSpeed = Util::dist(player.xSpeed, player.ySpeed);
    double alphaTarget = std::max(0.0, 255.0 - 600.0 * playerSpeed);
    foregroundAlpha_ += (alphaTarget - foregroundAlpha_) * frameLength / ((alphaTarget < foregroundAlpha_) ? 100.0 : 1500.0);

    if (Util::frameRandom(frameLength, 5.0)) {
        light.addMistParticle(player.x, player.y, player.color);
    }
    if (player.laser) {
        Laser* lase = player.laser.get();
        if (Util::frameRandom(frameLength, lase->length / 20.0)) {
            double mistPos = Util::randomDouble();
            light.addMistParticle(lase->startX + mistPos * lase->length * std::cos(lase->angle),
                                   lase->startY + mistPos * lase->length * std::sin(lase->angle),
                                   player.color);
        }
    }

    for (Nest* n : terrain.nestsTouchingRect(screenRect)) {
        n->updateVisuals(frameLength);
        if (!terrain.playerDamageCircles.empty()) {
            LaserTargetInfo laserInfo = player.laser ? player.laser->asLaserTargetInfo() : LaserTargetInfo{};
            for (auto& spec : n->applyDamageFromCircles(renderer_, terrain, laserInfo, player.x, player.y, player.rect)) {
                terrain.particles.spawnMiningParticles(10, n->color, spec.size, spec.x, spec.y);
            }
        }

        if (n->stage != n->maxStage) {
            double d = Util::dist(n->x - player.x, n->y - player.y);
            if (d < 300 && Util::frameRandom(frameLength, 0.2 + 0.5 * (300 - d) / 300.0)) {
                n->addEnemy(renderer_, terrain, player.x, player.y, player.rect);
            }
            for (int i = static_cast<int>(n->enemies.size()) - 1; i >= 0; --i) {
                Enemy* enemy = n->enemies[i];
                EnemyPlayerView epv{
                    player.x, player.y, player.xSpeed, player.ySpeed, player.immunityTimer, player.immunityTime,
                    player.rect, player.laser ? player.laser->asLaserTargetInfo() : LaserTargetInfo{},
                    [&](double dmg) { player.dealDamage(dmg); }
                };
                if (enemy->tick(frameLength, terrain, epv)) {
                    terrain.removeEnemy(enemy);
                    n->enemies.erase(n->enemies.begin() + i);
                }
            }
        } else {
            if (Util::frameRandom(frameLength, n->interactionDisplay.active ? 8.0 : 2.0)) {
                light.addMistParticle(n->x, n->y, n->color);
            }
        }
    }

    for (Cell* cell : terrain.cellsInRect(screenRect)) {
        if (cell->close(frame)) {
            LaserTargetInfo laserInfo = player.laser ? player.laser->asLaserTargetInfo() : LaserTargetInfo{};
            cell->tick(frameLength, terrain, laserInfo);
        }
    }

    terrain.displayManager.tick(frameLength, keysDown);

    light.tickEffects(frameLength);
    terrain.particles.tickParticles(frameLength);

    terrain.knockbackCircles = terrain.newKnockbackCircles;
    terrain.playerDamageCircles = terrain.newPlayerDamageCircles;

    return false;
}

// def draw_background(self, layer, window_size, frame): ...
// NOTE: uses Util::pyMod (matches Python's `%` operator), not std::fmod --
// left/top can be negative (world extends infinitely in x, and the
// camera can be above y=0), so sign-correct modulo matters here.
void World::drawBackground(SDL_Renderer* renderer, const Frame& frame) {
    const Asset& bg1 = GlobalAssets::getAsset("background_1");
    const Asset& bg2 = GlobalAssets::getAsset("background_2");
    // was: self.bg_width, self.bg_height = 3000, 3000 (originally the
    // PRE-SCALED destination size; now just the destination size we scale
    // to at draw time, native asset resolution notwithstanding).
    constexpr double kBgSize = 3000.0;

    double x1 = Util::pyMod(-frame.left * 1.0 * frame.zoom, kBgSize) / 2.0 - kBgSize / 2.0;
    double y1 = Util::pyMod(-frame.top * 1.0 * frame.zoom, kBgSize) / 2.0 - kBgSize / 2.0;
    Canvas::blit(renderer, bg1.texture, x1, y1, kBgSize, kBgSize);

    double x2 = Util::pyMod(-frame.left * 1.8 * frame.zoom, kBgSize) / 2.0 - kBgSize / 2.0;
    double y2 = Util::pyMod(-frame.top * 1.8 * frame.zoom, kBgSize) / 2.0 - kBgSize / 2.0;
    Canvas::blit(renderer, bg2.texture, x2, y2, kBgSize, kBgSize);
}

// def draw_foreground(self, layer, window_size, frame): ...
// NOTE: this is a PLAIN alpha blit (matches `layer.blit(self.foreground,
// (x,y))`, no special_flags) -- the MULT blending happens one level up,
// when the whole scratch layer this draws into gets composited onto the
// main layer (see drawWorld). foreground_alpha is applied here via
// SDL_SetTextureAlphaMod, matching `self.foreground.set_alpha(...)`.
void World::drawForeground(SDL_Renderer* renderer, const Frame& frame) {
    const Asset& fg = GlobalAssets::getAsset("foreground");
    constexpr double kFgSize = 10000.0;

    double x = Util::pyMod(-frame.left * 6.0 * frame.zoom, kFgSize) / 2.0 - kFgSize / 2.0;
    double y = Util::pyMod((-frame.top * 6.0 + 500.0) * frame.zoom, kFgSize) / 2.0 - kFgSize / 2.0;
    SDL_SetTextureAlphaMod(fg.texture, static_cast<uint8_t>(std::clamp(foregroundAlpha_, 0.0, 255.0)));
    Canvas::blit(renderer, fg.texture, x, y, kFgSize, kFgSize);
}

// def draw_world(self, window, window_size, frame, hitboxes=False, ...): ...
void World::drawWorld(SDL_Renderer* renderer, const Frame& frame, bool hitboxes,
                       bool kindVisibility, double tilt, bool crosshair) {
    if (layerW_ != frame.screenWidth || layerH_ != frame.screenHeight) {
        worldLayer_ = RenderTarget(renderer, frame.screenWidth, frame.screenHeight);
        scratchLayer_ = RenderTarget(renderer, frame.screenWidth, frame.screenHeight);
        finalLayer_ = RenderTarget(renderer, frame.screenWidth, frame.screenHeight);
        layerW_ = frame.screenWidth;
        layerH_ = frame.screenHeight;
    }

    worldLayer_.renderTo(renderer, [&] {
        if (kindVisibility) worldLayer_.clear({200, 200, 200, 255});
        else worldLayer_.clear({5, 5, 5, 255});

        light.drawGradient(renderer, frame, player.color, player.x, player.y, 600);
        if (player.laser && player.laser->collision.hit) {
            light.drawGradient(renderer, frame, player.color, player.laser->collision.x, player.laser->collision.y);
        }

        terrain.drawNestGradients(renderer, frame);
        terrain.drawEnemyGradients(renderer, frame);

        scratchLayer_.renderTo(renderer, [&] {
            scratchLayer_.clear({0, 0, 0, 0});
            drawBackground(renderer, frame);
        });
        Canvas::blit(renderer, scratchLayer_.texture(), 0, 0, frame.screenWidth, frame.screenHeight, BlendModes::rgbMult());

        light.drawGradient(renderer, frame, player.color, player.x, player.y);

        Color frameColor = terrain.getFrameColor(frame);
        scratchLayer_.renderTo(renderer, [&] {
            Canvas::rectFilled(renderer, Rect{0, 0, frame.screenWidth, frame.screenHeight}, frameColor);
        });
        Canvas::blit(renderer, scratchLayer_.texture(), 0, 0, frame.screenWidth, frame.screenHeight, BlendModes::rgbMult());

        light.drawEffects(renderer, frame);

        terrain.particles.drawPulseParticles(renderer, frame);

        player.draw(renderer, frame, hitboxes, tilt);

        terrain.drawCells(renderer, frame, hitboxes);
        terrain.drawEnemies(renderer, frame, hitboxes);
        terrain.particles.drawParticles(renderer, frame);
        terrain.drawNests(renderer, frame, hitboxes);
        terrain.drawTerrain(renderer, frame, hitboxes);

        int64_t nowMs = static_cast<int64_t>(Time::nowMs());
        terrain.drawHealthBars(renderer, frame, nowMs);
        terrain.drawInteractionDisplays(renderer, frame, nowMs);

        if (!kindVisibility) {
            scratchLayer_.renderTo(renderer, [&] {
                scratchLayer_.clear({255, 255, 255, 255});
                drawForeground(renderer, frame);
                light.drawThickGradient(renderer, frame, player.x, player.y);
                if (player.laser && player.laser->collision.hit) {
                    light.drawThickGradient(renderer, frame, player.laser->collision.x, player.laser->collision.y);
                }
            });
            // was: special_flags=pygame.BLEND_MULT (the plain/un-prefixed
            // flag, which multiplies ALL channels including alpha) --
            // rgbaMult(), not rgbMult().
            Canvas::blit(renderer, scratchLayer_.texture(), 0, 0, frame.screenWidth, frame.screenHeight, BlendModes::rgbaMult());
        }

        if (crosshair) {
            // was: pygame.draw.line(layer, (100,100,100,0.3), ...) -- the
            // alpha component (0.3) is a float where pygame expects an
            // int 0-255, AND this world_layer surface was created without
            // pygame.SRCALPHA in the Python (`pygame.Surface(real_window_size)`,
            // no per-pixel alpha support), so the alpha component would
            // most likely be ignored entirely by pygame -- resulting in a
            // fully OPAQUE gray line despite the apparent "faint" intent.
            // Implemented as opaque gray here to match that most-likely
            // actual behavior, flagged since this is inference about an
            // ambiguous original, not a certainty.
            Canvas::line(renderer, {frame.screenWidth * 0.45, frame.screenHeight / 2.0},
                         {frame.screenWidth * 0.55, frame.screenHeight / 2.0}, Color{100, 100, 100, 255}, 2);
            Canvas::line(renderer, {frame.screenWidth / 2.0, frame.screenHeight * 0.45},
                         {frame.screenWidth / 2.0, frame.screenHeight * 0.55}, Color{100, 100, 100, 255}, 2);
        }
    });

    // was: layer, cx, cy = rotate_and_get_offset(layer, real_window_size[0]/2,
    // real_window_size[1]/2, math.radians(tilt)); layer.blit(layer, (cx, cy))
    // -- pivot IS the screen center here, so this is the simple "rotate
    // around own center" case (blitRotated's default), same as
    // LaserImpact. Baked into a SEPARATE offscreen buffer (not just
    // rotated at final-blit time) specifically so bloom (below) samples
    // the POST-rotation frame, matching the Python's actual execution
    // order (tilt runs before bloom, both applied to the same `layer`).
    RenderTarget* composedLayer = &worldLayer_;
    if (tilt != 0) {
        finalLayer_.renderTo(renderer, [&] {
            finalLayer_.clear({0, 0, 0, 0});
            // V (pygame degrees) = degrees(radians(tilt)) == tilt; angle_SDL = -V = -tilt
            Canvas::blitRotated(renderer, worldLayer_.texture(), frame.screenWidth / 2.0, frame.screenHeight / 2.0,
                                 frame.screenWidth, frame.screenHeight, -tilt);
        });
        composedLayer = &finalLayer_;
    }

    RenderTarget& bloomResult = bloom_.getBloom(renderer, composedLayer->texture(), frame.screenWidth, frame.screenHeight);
    composedLayer->renderTo(renderer, [&] {
        Canvas::blit(renderer, bloomResult.texture(), 0, 0, frame.screenWidth, frame.screenHeight, BlendModes::rgbAdd());
    });

    Canvas::blit(renderer, composedLayer->texture(), 0, 0, frame.screenWidth, frame.screenHeight);

    if (crosshair) {
        // was: the SECOND crosshair draw, AFTER the tilt-rotation step --
        // stays screen-fixed regardless of tilt, unlike the first
        // (gray) crosshair drawn inside worldLayer_ above.
        double size = 10.0;
        Canvas::line(renderer, {frame.screenWidth / 2.0 - size, frame.screenHeight / 2.0},
                     {frame.screenWidth / 2.0 + size, frame.screenHeight / 2.0}, Color{255, 0, 0, 255}, 2);
        Canvas::line(renderer, {frame.screenWidth / 2.0, frame.screenHeight / 2.0 - size},
                     {frame.screenWidth / 2.0, frame.screenHeight / 2.0 + size}, Color{255, 0, 0, 255}, 2);
    }
}
